### PR TITLE
SideNav & TopNav upgrades + end-to-end testing

### PR DESCRIPTION
**Pull request by AniaMakes**
## Description - [Trello ticket 497-2](https://trello.com/c/9l0OD5Qr/497-top-nav-dropdown-refactoring)

Refactored the isExpanded logic so that it now works as ticket requires. 
Also added shortid module to solve "multiple children with same ID" warning in the console. 

- [x] SideNav dropdowns should remain expanded if the active page is within that drop down
- [x] TopNav - if the active page is in a dropdown, then the dropdown select button should be greyed out (i.e. use styles for "active" state)
- [x] e2e tests for Dropdown - for both TopNav (desktop) and SideNav (mobile)

## Review template

- [ ] checked that this PR resolves the spec provided from trello / this description
- [ ] ensured that this PR does not introduce any obvious bugs

### CODE REVIEW

File: src/components/Header/SideNav/Dropdown.js [hunk 1]
```diff
@@ -59,8 +60,17 @@ export default class Dropdown extends PureComponent {
   }
 
   render() {
-    const { items, children } = this.props
-    const { isExpanded } = this.state
+    const { items, children, path } = this.props
+    let { isExpanded } = this.state
+
+    const isServicePage = path.search(servicesRegExp) > -1
```
Comment (omonk) [inline src/components/Header/SideNav/Dropdown.js]: I think this is already written somewhere else. In utils?

File: src/components/Service/getSections.js [hunk 2]
```diff
@@ -7,7 +9,7 @@ export default workStage =>
       // arrays start with 0 so that's why the + 1
       const getCurrent = part => `section${capitalize(part)}${index + 1}`
       return {
-        id: workStage.id,
+        id: `${workStage.id}-${generate()}`,
```
Comment (omonk) [inline src/components/Service/getSections.js]: You don't need workStage at all here. Generate will suffice

File: src/components/Header/SideNav/Dropdown.js [hunk 3]
```diff
@@ -59,8 +60,17 @@ export default class Dropdown extends PureComponent {
...
+    const serviceTitle = isServicePage ? path.match(servicesRegExp)[0] : null
+    const isSpecialityPage = path.includes('speciality')
+    const service = isSpecialityPage ? getService(path) : serviceTitle
+
+    isExpanded =
+      isExpanded ||
+      items.filter(item => item.to.startsWith(`/${service}`)).length > 0
```
Comment (omonk) [inline src/components/Header/SideNav/Dropdown.js]: Could this lead to not being able to close a dropdown?

Comment (AniaMakes) [reply src/components/Header/SideNav/Dropdown.js]: ugh, yes, good catch.

File: src/components/Service/getSections.js (hunk 2 above)
Comment (AniaMakes) [reply src/components/Service/getSections.js]: I wondered about that.

File: src/components/Header/SideNav/Dropdown.js (hunk 1 above)
Comment (GBouffard) [reply src/components/Header/SideNav/Dropdown.js]: Good call. I added that 2 places for a previous ticket (TopNav components) but that wasn't part of a utils file. Now that it's used here as well it makes sense to move it to a utils file at the /Header level

Comment (GBouffard) [reply src/components/Header/SideNav/Dropdown.js]: I moved it 👍

File: src/components/Header/SideNav/Dropdown.js (hunk 3 above)
Comment (omonk) [reply src/components/Header/SideNav/Dropdown.js]: Also I didn't notice this yesterday but you're referencing state and changing its value in the render without setting the state of the actual component which is not great. React has component lifecycle methods for what we're trying to do here

File: tests/e2e/sideNav-dropdown-test.js [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+})
+
+test('a hamburger is present and opens the side nav', async t => {
+  const hamburger = await Selector('[class^="Hamburger"').filterVisible()
+  await t.expect(hamburger.exists).ok()
+  await t.click(hamburger)
+
+  const sideNav = await Selector('[class^="SideNav"]').filterVisible()
+  await t.expect(sideNav.exists).ok()
+})
```
Comment (omonk) [inline tests/e2e/sideNav-dropdown-test.js]: So the main reason I wanted to get some tests for the sidenav was so we could catch any functionality errors. 

Lets add another one

`a dropdown opens when the dropdown title is clicked`

Comment (GBouffard) [reply tests/e2e/sideNav-dropdown-test.js]: We have plenty of tests now:
 ✓ we are on the homepage
 ✓ a hamburger is present on the page
 ✓ the side nav is not expanded at its initial state
 ✓ when the hamburger button is clicked the side nav expands
 ✓ clicking a side Nav link redirects to the correct url and closes the side Nav
 ✓ when the side nav expands for the first time no sub-item is shown by default
 ✓ when an item is clicked its sub-items appear
 ✓ clicking a sub-item redirects to the correct url and closes the side Nav
 ✓ After a sub-item has been clicked and if the sideNav is re-opened, the subitems are still visible and the selected one highlighted

PS: We looked at testcafe-react-selectors but in the end, it was easier and more convenient to use the API made available by Testcafe.

File: tests/e2e/sideNav-dropdown-test.js [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+
+  const joinUsLink = sideNavPanel.find('a').withText('Join us')
+  await t.click(joinUsLink)
+
+  const location = await getWindowLocation()
+  await t.expect(location.href).contains(`${baseUrl}/join-us`)
+  await t.expect(sideNavPanel.exists).notOk()
+})
+
+test('when the side nav expands for the first time no sub-item is shown by default', async t => {
```
Comment (omonk) [inline tests/e2e/sideNav-dropdown-test.js]: Don't we auto expand dropdowns if the active page is within the dropdown?

**Comment by omonk** [comment]
super thorough tests! nice one guys

File: tests/e2e/sideNav-dropdown-test.js (hunk 5 above)
Comment (AniaMakes) [reply tests/e2e/sideNav-dropdown-test.js]: Yes, but this example starts at `baseUrl` (line 34).

Comment (omonk) [reply tests/e2e/sideNav-dropdown-test.js]: I was more referring to the name of the test. If anyone comes to the project should be able to garner a bit information on behaviour from tests. The test name should reflect that we don't open any dropdowns if none of the dropdowns contain links to the current path, we should also have a test to reflect the opposite

**Comment by omonk** [comment]
rebased and merged in https://github.com/yldio/yld.io/pull/214

**Pull request closed without merging**